proposals/lip-0045.md

```diff
@@ -779,199 +743,193 @@ The `apply` function applies a CCM. It is specified in [LIP 0049][lip-0049].
+#### bounce and bounceWithoutFee
 
-##### Parameters
-
-This function has the following input parameters in the order given below:
-
-* `feeAddress`: The address of the account paying the fee.
-* `moduleID`: The ID of the module calling this function.
-* `crossChainCommandID`: The ID of the command calling this function.
-* `receivingChainID`: The ID of the receiving chain.
-* `fee`: The message fee.
-* `status`: The message status.
-* `params`: The message parameters.
-
-##### Returns
-
-A boolean indicating the success of the function execution.
-
-##### Execution
+The `bounce` and `bounceWithoutFee` functions return a CCM that errored during the CCM execution stages to the sending chain. 
+`bounceWithoutFee` is called if an error occurs during the cross-chain command execution (see [`apply` function][lip-0049#apply]) in the receiving chain. 
+In this case, the fee property is set to 0 and the message is returned only if the fee was larger than a minimum return fee and the status of the CCM is `CCM_OK`.
+`bounce` is called if an error occurs in any other stage.
+In this case, the fee is not changed and the message is returned if the status of the CCM is `CCM_OK`.
 
 ```python
-sendInternal(feeAddress, moduleID, crossChainCommandID, receivingChainID, fee, status, params):
-    # Processing on the mainchain
-    if ownChainAccount.ID == MAINCHAIN_ID:
-        # Chain has to exist on mainchain
-        if chainAccount(receivingChainID) does not exists:
-            return False
-        partnerChainID = receivingChainID
-    # Processing on a sidechain
-    else:
-        # Check for direct channel
-        if chainAccount(receivingChainID) does not exists:
-            partnerChainID = mainchainID
-        else:
-            partnerChainID = receivingChainID
-
-    # Chain must be live; This checks is always on the receivingChainID
-    if not isLive(receivingChainID):
-        return False
-    # Chain status must be active
-    if not chainAccount(partnerChainID).status == CHAIN_ACTIVE:
-        return False
-
-    # Create cross-chain message
-    ccm = {
-        "nonce": ownChainAccount.nonce,
-        "moduleID": moduleID,
-        "crossChainCommandID": crossChainCommandID,
-        "sendingChainID": ownChainAccount.ID,
-        "receivingChainID": receivingChainID,
-        "fee": fee,
-        "status": status,
-        "params": params
-    }
-
-    if validateFormat(ccm) == False:
-        return False
-
-    for each module mdl for which beforeSendCCM exists:
-        mdl.beforeSendCCM(feeAddress, ccm)
-        if the previous call fails:
-            return False
-
-    addToOutbox(partnerChainID, ccm)
-    ownChainAccount.nonce += 1
+def bounce(ccm: CCM, errorCode: uint32) -> None:
+    ccmID = sha256(encode(crossChainMessageSchema, ccm))
+    emitEvent(
+        module = MODULE_NAME_INTEROPERABILITY,
+        name = EVENT_NAME_CCM_PROCESSED,
```

Comment:
I find this event overall a bit confusing and we also do not use it consistently. Here, for instance, we use it both to emit an event about the kind of error that happened (which at the same time is the status of the returned message) as well as to say what happens further with the CCM. In some cases (`ERROR_CODE_INVALID_FORMAT`), we emit only an event about the error, but do not say what happens with the CCM. I wonder if it would not be sufficient to just emit one event in all cases.

Reply:
What I find further confusing is that we have constants from 0 (`CCM_OKAY`) to 11 (`ERROR_CODE_MESSAGE_FEE_ERROR`) that refer to very different aspects of the CCM processing. For instance:

- `CCM_OK` is used both as default `status` code when sending a message (basically just meaning, this is not a return error CCM) and to signal that a CCM was executed successfully, which are quite different things.
- `CCM_APPLIED` is used for a CCM processed event in case of an error although it suggests that it means that a CCM is successfully applied?
- We have `ERROR_CODE_INVALID_CCM` and `ERROR_CODE_FAILED_CCM`, which seem rather generic, but then there are more specific errors for the case of invalid and failing CCMs.

Reply:
Maybe the event could have the properties:

- "executionResult" in {"success", "failed", "invalid"} 
- "action" in {"apply", "bounce", "forward", "discard"}
- "errorCode" in { error codes already defined for different failures}

Of course property and value names could be different, in particular, or values could also be integers.

Note that any message should either be successful, failing or invalid, and the action of the interoperability module should always be either apply (in case the CCM is successful), bounce (this can happen for failing or invalid CCMs), forward or discard.

Reply:
Regarding the first point: it was a copy/paste mistake, when we apply we use `CCM_APPLIED`. 

I would reserve 'invalid' and 'failed' for the error that you get during execution stack. if the channel is unavailable, it makes sense to use a different error code. 

Regarding your proposal, I still think there are some unclear points: for instace, what would be the executionResult in case of action=forwarding?

Reply:
addressed together

Reply:
Looks good now.